lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/TaskMessageObserver.cs

```diff
@@ -0,0 +1,134 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Org.Apache.REEF.Network.Group.Driver.Impl;
+using Org.Apache.REEF.Network.NetworkService;
+using Org.Apache.REEF.Wake.Remote;
+using Org.Apache.REEF.Wake.Remote.Impl;
+
+namespace Org.Apache.REEF.Network.Group.Task.Impl
+{
+    /// <summary>
+    /// The observer for a Task that multiplexes to the node observers associated with that Task.
+    /// </summary>
+    internal sealed class TaskMessageObserver :
+        IObserver<NsMessage<GeneralGroupCommunicationMessage>>, 
+        IObserver<IRemoteMessage<NsMessage<GeneralGroupCommunicationMessage>>>
+    {
+        private readonly Dictionary<NodeObserverIdentifier, IObserver<NsMessage<GeneralGroupCommunicationMessage>>> _observers =
+            new Dictionary<NodeObserverIdentifier, IObserver<NsMessage<GeneralGroupCommunicationMessage>>>();
+
+        private readonly StreamingNetworkService<GeneralGroupCommunicationMessage> _networkService;
+        private readonly object _registrationLock = new object();
+        private bool _hasRegistered = false;
+        private volatile NsMessage<GeneralGroupCommunicationMessage> _registrationMessage;
+
+        public TaskMessageObserver(StreamingNetworkService<GeneralGroupCommunicationMessage> networkService)
+        {
+            _networkService = networkService;
+        }
+
+        /// <summary>
+        /// Registers a node associated with the Task.
+        /// </summary>
+        public void RegisterNodeObserver<T>(NodeMessageObserver<T> observer)
+        {
+            _observers.Add(NodeObserverIdentifier.FromObserver(observer), observer);
+        }
+
+        /// <summary>
+        /// This is called directly from the observer container with the registered IPEndpoint
+        /// of the Task ID.
+        /// </summary>
+        public void OnNext(NsMessage<GeneralGroupCommunicationMessage> value)
+        {
+            Handle(value);
+        }
+
+        /// <summary>
+        /// Handles the group communication message.
+        /// </summary>
+        private void Handle(NsMessage<GeneralGroupCommunicationMessage> value, bool isRegistration = false)
+        {
+            // This is mainly used to handle the case should ObserverContainer
+            // decide to trigger handlers concurrently for a single message.
+            if (isRegistration)
+            {
+                // Process the registration message
+                _registrationMessage = value;
+            }
+            else if (_registrationMessage != null && value == _registrationMessage)
+            {
+                // This means that we've already processed the message.
+                // Ignore this message and discard the reference.
+                _registrationMessage = null;
```

Comment:
To release the reference such that `_registrationMessage` can be garbage collected if needed.